README.md
# PSJsonCredential

[![PSGallery Version](https://img.shields.io/powershellgallery/v/PSJsonCredential.png?style=for-the-badge&logo=powershell&label=PowerShell%20Gallery)](https://www.powershellgallery.com/packages/PSJsonCredential/) [![PSGallery Downloads](https://img.shields.io/powershellgallery/dt/PSJsonCredential.png?style=for-the-badge&label=Downloads)](https://www.powershellgallery.com/packages/PSJsonCredential/)

This PowerShell module contains a set of functions for storing a PSCredential object in a JSON file. Previous versions of this module relied on the CryptoAPI which limited it to Windows platforms only. This version uses a user-defined key to convert the password to a secure string that can be stored in the JSON file. The module should now work in Windows PowerShell and PowerShell 7.

## Release

The most current version is available from the PowerShell Gallery.

```powershell
PS C:\> Install-Module PSJsonCredential
```

## Why?

You can achieve similar results using the [Export-CliXML](http://go.microsoft.com/fwlink/?LinkId=821767) and [Import-Clixml](http://go.microsoft.com/fwlink/?LinkId=821813) cmdlets. But those techniques rely on the CryptoAPI and are limited to Windows platforms. The files can also only be used on the computer where they were created only by the original user. Earlier versions of this module relied on these techniques.

The current version of the module relies on a user-defined key which you can think of as a password or passphrase. __You need to know the key to export and import the credential__. This has the added benefit of making the JSON file portable between computers.

## Usage

### Export

You can pipe any PSCredential object to [Export-PSCredentialToJson](docs/Export-PSCredentialToJson.md).

```powershell
PS C:\> $cred = Get-Credential Company\Administrator
PS C:\> $skey = Read-Host "Enter a 16 character key" -asSecureString
Enter a 16 character key: ****************
PS C:\> $skey.Length
16
PS C:\> $cred | Export-PSCredentialToJson -Path c:\work\admin.json -key $skey
```

The export process will also capture metadata information about who converted the credential.

```powershell
PS C:\> Get-Content C:\work\admin.json
{
  "UserName": "company\\administrator",
  "Password": {
    "value": "76492d1116743f0423413b16050a5345MgB8AEUARwBrAHAASABwAE8AdgBOAEgAWgA2AHkAWAA4AEYANgA4AEkAVQBKAEEAPQA9AHwAZQAzADAAMAA1ADEAOQAzADEANAA0AGIAYQA3AGEAOQBmAGMAZQAwADQANAAzADMAOAAxADEAMgA5ADAAMABkADkANwAzADAAZgAzADcAYgA0AGYAZQBiAGUANQBhADAAMgBmADEAZABkAGUAZQBjADMAZAA2AGYAYQA5AGUAMQA="
    },
  "Metadata": {
    "ExportDate": "7/19/2023 7:54:02 PM",
    "ExportUser": "DESK10\\Jeff",
    "ExportComputer": "DESK10"
    "Comment": ""
  }
}
```

Beginning with v2.2.0, you also have the option to encrypt the user name in the JSON file.

```powershell
PS C:\> $credential | Export-PSCredentialToJson -NoMetaData -EncryptUserName -Key $key -Path D:\Save\export.json
PS C:\> Get-Content D:\Save\export.json
{
    "Username":  "76492d1116743f0423413b16050a5345MgB8AHUASwBjADQAagArAFgAUABaAGoARwB0AGcASgBJAFcANgBqADkAcABWAGcAPQA9AHwAOQBlADkAZQA5AGMAZAA1ADgAOQBlAGIAYgBlADgAYQA5ADYAMwBhADIANQAwADQAOQA2AGMAMgBjADEAZQA1ADAAZAAzAGMAMABjADAAMgAwAGQAZgA3ADQANwBlADAANwA4ADcAOQBkADMAOAA0AGMANQA2ADEAYQBiADYAZABhAGUAYwAyAGQAYQA2ADUAMABhADkAMQAzADkAMgA0AGQAMwBmADEANABlADUAYQA3ADYAZgA4AGIAOQAzADUA",
    "Password":  "76492d1116743f0423413b16050a5345MgB8ADQAcABVAEMAaQA5AHQARgBTAGMAYQBZAG0ATQBwAHAAUwByADIASABXAEEAPQA9AHwAZQAyADUAOAAxADgAYwAyADEAOQBlAGMANgAyAGYAYgAyADYAMwA3AGEAYwAxADYANgA0AGQAZABkADMAMQBmADYANQBiADgANAAwADEAZgA1AGMAOQBhAGYAMQBiADEAYwAyADMAYwAwAGEAOQBhAGMAZABjADIAOAA2ADMAMQA="
}
```

### Viewing the Export

You can "get" the credential from the JSON file but without converting the password with [Get-PSCredentialFromJson](docs/Get-PSCredentialFromJson.md)

```powershell
PS C:\> Get-PSCredentialFromJson -Path C:\work\admin.json


   Path: C:\work\admin.json


Username       : company\administrator
Password       : 76492d1116743f0423413b16050a5345MgB8AEwAcAB3AGoARABOAGk
                 AbABKAFgARAB6AEEAUQBPADYAVwBPAHgAZgB1AEEAPQA9AHwAYwA0AG
                 MAZgAyAGMAYQBjAGEAMQBlAGMAMQBhADgAOQAwADAANgBkADgAYgA3A
                 DMANgA4AGUAZABiADUAOAA0AGYANgA3ADYAYgBlAGYAOAA0AGUAOQBl
                 ADIAOAAwADcANQBjADgAYwA1AGUAMQBhAGMANgBhADAAYgBkADIAYgA=
ExportDate     : 8/2/2023 12:53:32 PM
ExportUser     : COMPANY\Jeff
ExportComputer : DESK11
Comment        :
```

There is also a named list view called `nometa`.

```powershell
PS C:\> Get-PSCredentialFromJson -Path C:\work\admin.json | Format-List -view nometa


   Path: C:\work\admin.json


Username       : company\administrator
Password       : 76492d1116743f0423413b16050a5345MgB8AEwAcAB3AGoARABOAGk
                 AbABKAFgARAB6AEEAUQBPADYAVwBPAHgAZgB1AEEAPQA9AHwAYwA0AG
                 MAZgAyAGMAYQBjAGEAMQBlAGMAMQBhADgAOQAwADAANgBkADgAYgA3A
                 DMANgA4AGUAZABiADUAOAA0AGYANgA3ADYAYgBlAGYAOAA0AGUAOQBl
                 ADIAOAAwADcANQBjADgAYwA1AGUAMQBhAGMANgBhADAAYgBkADIAYgA=
Comment        :
```

### Importing

And when you are ready, you can import the credential using [Import-PSCredentialFromJson](docs/Import-PSCredentialFromJson.md)

```powershell
PS C:\> $in = Import-PSCredentialFromJson -Path C:\scripts\admin.json -key $skey --Verbose
VERBOSE: [BEGIN  ] Starting: Import-PSCredentialFromJson
VERBOSE: [PROCESS] Processing credential from C:\scripts\admin.json
VERBOSE: [PROCESS] Preparing key with a length of 16
VERBOSE: [PROCESS] Converting to System.Security.SecureString
VERBOSE: [PROCESS] Creating credential for company\administrator
VERBOSE: [END    ] Ending: Import-PSCredentialFromJson
PS C:\> $in

UserName                                  Password
--------                                  --------
Company\Administrator System.Security.SecureString


PS C:\> $in.GetNetworkCredential().password
P@ssw0rd
```

## Samples

You can find demonstration scripts using the module commands in the [Samples](samples) folder.

## Is It Safe

The commands use the secure string convert cmdlets. The password is encoded with a user-defined key which can be a password or phrase of length 16, 24, or 32. Any password stored on disk is a potential security risk.

Because the JSON file is a plain text file, the user and computer name will be visible. You should still take precautions to secure and protect the JSON file.

### Important

Storing any credential to disk poses a potential security risk. It is up to you to decide if you wish to use the commands in this module. Some organizations may have security or usage policies in place that forbid the behavior.

> *__Use with caution and at your own risk.__*


Changelog.md
# Change Log for PSJsonCredential

## v2.2.0

### Added

- Added a parameter to `Export-PSCredentialToJson` to encrypt the credential user name.
- Added a `-Comment` parameter to `Export-PSCredentialToJson` which will include it in metadata. This change results in new parameter sets.
- Added format file `psjsoncredential.format.ps1xml` with a default list view. There is also a named list view called `nometa`.

### Fixed

- Revised private function to convert secure strings to better handle differences between PowerShell 7 and Windows PowerShell. ([Issue #5](https://github.com/jdhitsolutions/PSJsonCredential/issues/5))

### Changed

- Restructured module layout.
- Modified import and export commands to accept a secure string for the key and not plain text password. **This is a breaking change**
- Moved .NET methods to private functions for future Pester tests.
- Changed output from `Get-PSCredentialFromJson` to include the metadata comment.
- Added sample files.
- Help updates.
- General code cleanup.
- Revised Pester tests for v5.
- Updated `README.md`.

## v2.1.0

- fixed metadata bug on non-Windows platforms ([Issue #4](https://github.com/jdhitsolutions/PSJsonCredential/issues/4))
- Minor Pester test correction

## v2.0.0

- Updated manifest to support Desktop and Core PSEditions
- Modified commands to use a Key to secure password and not rely on the Crypto API
- Made metadata optional on `Export-PSCredentialToJson`
- Updated help
- Updated `README.md`
- Updated Pester tests
- Removed online help download links

## v1.3.1

- file cleanup for the PowerShell Gallery
- Updated `README.md`

## v1.3.0

- Added parameter validation to `Path` in `Export-PSCredentialToJson` for .json extension.
- Fixed a new bug with an apparent change in `ConvertFrom-Json`. Password value was no longer a simple string.
- Code cleanup

## v1.2.0

- code cleanup
- revised license
- fixed Pester test bugs
- Updated `README.md`
- Updated help

## v1.1.3.0

- changed author in manifest
- changed to semantic version numbers

## v1.1.2

- Added links for online help
- Updated help documentation
- updated version number

## v1.1.1

- revised help

## v1.1.0

- Added path to output from `Get-PSJsonCredential` ([Issue #1](https://github.com/jdhitsolutions/PSJsonCredential/issues/1))
- Updated documentation

## v1.0.0

- public release to PSGallery

## v0.1.0

- updated manifest
- updated README
- created help markdown documents
- created external help

## v0.0.9

- initial build


functions/Export-PSCredentialToJson.ps1
Function Export-PSCredentialToJson {
    [CmdletBinding(SupportsShouldProcess,DefaultParameterSetName = "comment")]
    [OutputType('None', 'System.IO.FileInfo')]

    Param(
        [Parameter(
            Position = 0,
            Mandatory,
            HelpMessage = 'Enter the name and path of the json file to create.'
        )]
        [ValidateNotNullOrEmpty()]
        [ValidatePattern('\.json$')]
        [String]$Path,

        [Parameter(Mandatory, ValueFromPipeline)]
        [ValidateNotNullOrEmpty()]
        [PSCredential]$Credential,

        [Parameter(Mandatory, HelpMessage = 'Enter a key password or passphrase of length 16, 24 or 32.')]
        [ValidateScript({ $_.length -eq 16 -OR $_.length -eq 24 -OR $_.length -eq 32 })]
        [System.Security.SecureString]$Key,

        [Parameter(HelpMessage = 'Encrypt the user name as a secure string')]
        [Switch]$EncryptUserName,

        [Parameter(HelpMessage = "Don't overwrite an existing file.")]
        [Switch]$NoClobber,

        [Parameter(ParameterSetName = 'noMetadata', HelpMessage = 'Do not include metadata in the json file')]
        [Switch]$NoMetadata,

        [Parameter(ParameterSetName = 'metadata')]
        [string]$Comment,

        [Switch]$PassThru
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.MyCommand)"
        Write-Verbose "[BEGIN  ] Exporting credential to $Path"
    } #begin

    Process {
        Write-Verbose "[PROCESS] Processing credential for $($credential.UserName)"

        if ($NoClobber -AND (Test-Path -Path $path)) {
            Write-Warning "$path already exists and NoClobber specified."
            #bail out out
            Return
        }

        Write-Verbose "[PROCESS] Preparing key with a length of $($key.length)"
        $cKey = _convertSecureString $key
        $secret = _getBytes $cKey

        Try {
            #create a custom object from the credential and convert it to JSON
            if ($EncryptUserName) {
                $content = $Credential | Select-Object -Property @{Name = 'Username';
                    Expression = {
                        #convert to a secure string then convert from using the secret
                        ConvertTo-SecureString -String $_.UserName -AsPlainText -Force |
                        ConvertFrom-SecureString -Key $secret
                    }
                },
                @{Name = 'Password'; Expression = { $_.password | ConvertFrom-SecureString -Key $secret } }
            }
            else {
                $content = $Credential | Select-Object -Property Username, @{Name = 'Password';
                    Expression = { $_.password | ConvertFrom-SecureString -Key $secret }
                }
            }
            if ( -not $NoMetadata) {
                $meta = [PSCustomObject]@{
                    ExportDate     = (Get-Date).ToString()
                    ExportUser     = "$([system.environment]::UserDomainName)\$([system.environment]::username)"
                    ExportComputer = [System.Environment]::MachineName
                    Comment        = $Comment
                }
                $content | Add-Member -MemberType NoteProperty -Name Metadata -Value $meta
            } #if -not metadata
            else {
                Write-Verbose '[PROCESS] Excluding metadata'
            }

            $json = $content | ConvertTo-Json -ErrorAction Stop

            if ($PSCmdlet.ShouldProcess($path, 'Create json Credential File')) {
                $json | Set-Content -Path $Path -ErrorAction Stop

                if ($PassThru ) {
                    Get-Item -Path $path
                }
            } #WhatIf
        } #Try
        Catch {
            Throw $_
        } #Catch
    } #process

    End {
        Write-Verbose "[END    ] Ending: $($MyInvocation.MyCommand)"
    } #end

}
